File: 2024/Day02/day02.kt
import java.io.File
import kotlin.math.abs 

fun main() {
    val lines: List<String> = File("input.txt").readLines()

    val lists = lines.map {it.split(" ").map{ it.toInt()} }

    println(part1(lists))
    println(part2(lists))
}

fun part1(lists : List<List<Int>>) : Int {

    val res = lists.map {
        val dir = if (it[0] > it[1]) 1 else -1
        for (i in 0 until it.size-1) {
            val diff = dir * (it[i] - it[i+1])
            if (diff > 3 || diff < 1){
                return@map 0
            }
        }
        1
    }.sum()
    
    return res

} 

// This is very bad...
fun part2(lists : List<List<Int>>) : Int {

    val res = lists.sumOf {
        var sum = 0
        for (i in it.indices) {
            var ok = false
            val listWithout = it.filterIndexed {idx, x -> idx != i}
            val isOk = part1(listOf(listWithout)) == 1
            //println("$listWithout $isOk")
            ok = ok || isOk

            if (ok) return@sumOf 1
            0
        }
        sum
    }
    
    return res

}
// kotlinc day02.kt -include-runtime -d day02.jar
// java -jar day02.jar

File: 2025/day02/day02.py
intervals = [
        tuple(intervals.split("-"))
        for intervals in open('input.txt').readlines()[0].split(",")
]

def part1(pairs):
    s = 0
    for a, b in pairs:
        for x in range(int(a), int(b) + 1):
            xs = str(x)
            if len(xs) % 2 == 1:
                continue
            half = xs[:len(xs)//2]
            if x == int(half * 2):
                s += x
    return s

def part2(pairs):
    s = 0
    for a, b in pairs:
        for x in range(int(a), int(b) + 1):
            xs = str(x)
            for i in range(2, len(xs) + 1):
                if len(xs) % i != 0: continue
                part = xs[:len(xs)//i]
                if x == int(part * i):
                    s += x
                    break
    return s

print(part1(intervals))
print(part2(intervals))


